### DISCUSSION TITLE
API versioning

### DISCUSSION THREAD

**Discussion by baptisteArno** [supabase/supabase#2957] [Questions] [answered]
I'm wondering, if I'm making a breaking change in my DB schema:

How to avoid having a client consuming the old schema?
Is there any way to version Supabase?

**Comment by dshukertjr** [comment]
Curious to hear how other people do it, but I think utilizing views and postgres functions for versioning might be one route. I have not yet hit a point where I have to introduce a breaking change in my DB schema in my app, but I plan to create a v2 view, functions when that happens!

**Reply by silentworks** [reply]
I like this strategy except that I would flip it and create v1 views, functions instead and make the default schema be the new version.

**Comment by steve-chavez** [comment, marked as answer]
Versioning can be done with schemas that contain views + functions. You can name your schema `v1` and a new one with breaking changes `v2`.

The `postgrest-js` library has a [schema parameter](https://github.com/supabase/postgrest-js/blob/master/src/PostgrestClient.ts#L9) that allows switching the schema on the fly.

To do the above, you need to whitelist the exposed schemas on the Supabase dashboard - this feature is being worked right now.

---

Note that you can avoid breaking changes by using views in some cases. This is a simple example:

```sql
-- suppose you have this table
create table v1.items(
  id int primary key,
  name text,
  stock int
);

-- and you drop the stock column and decide to move this data to other table
alter table v1.items drop column stock;

create table private.items_details(
  item_id int,
  stock int
);

-- you can avoid a breaking change by moving the previous table to a private schema
alter table v1.items set schema private;

-- and creatng a view that has the stock column
create view v1.items as
select it.id, it.name, itd.stock -- the stock column is back again
from private.items it
join private.items_details itd on ..

-- the previous statements can be executed on a single transaction, so it would mean zero downtime for your users
```
